### ISSUE TITLE
Event framework

### ISSUE THREAD

**Issue by kshlm** [gluster/glusterd2#132]
GD2 needs a well defined and easy to use events framework.

To allow GD2 to be flexible and as small as possible, we'd like to use events to get particular features implemented outside the core of GD2, for eg. management of SHD, quotad, snapd and other daemons.

**Comment by atinmu** [comment]
@nnDarshan Would you want to work on this?

**Comment by aravindavk** [comment]
This issue is for internal consumers, created this issue for providing hooks for other components. 
#370 

@kshlm if you have any plans to provide this feature otherthan middleware approach, let me know.

**Comment by nnDarshan** [comment]
> @nnDarshan Would you want to work on this?

will work on https://github.com/gluster/glusterd2/issues/418

**Comment by kshlm** [comment]
I'm picking this up myself.

**Comment by kshlm** [comment]
I've begun working on this. The expected API of the events package is described below.

```golang
// EventHandler is a function that is registered to be called when an event happens
// event is the name of the event that happened and data is any extra data that was attached to the event as a json document
type EventHandler func(event string, data string)

// Register an event handler function to be called when the given events happen
// If no events are specified, handler function is called for all events
func Register(f EventHandler, events ...string) error

// Event represents an event in GD2
type Event struct {
  // Name is the the name of the event
  Name string
  // Data is any additional data attached to the event. Should be a JSON document
  Data string
  // Global should be set to true to broadcast event to the full GD2 cluster. If not event is only broadcast in the local node
  Global bool
}

// Broadcast broadcasts events to all registered event handlers
func Broadcast(e Event) error
```
I'm currently trying to find an efficient broadcast package. A simple broadcast method could be implemented using channels, but it doesn't

**Comment by aravindavk** [comment]
> I'm currently trying to find an efficient broadcast package. A simple broadcast method could be implemented using channels, but it doesn't

How about reusing Transaction framework itself? On event spawn a goroutine to run a OnEvent transaction to run in all nodes of Cluster. OnEvent function gets list of registered consumer for that event and executes/runs it locally.

**Comment by kshlm** [comment]
An internal events framework is implemented by #474. Further enhancements to enable external consumers, and to add events will be done in seperate PRs.
